### PR TITLE
Default preprocessing transforms handles diagonalization of measurements

### PR DESCRIPTION
**Pull request by astralcai**
The original intention was to include it in https://github.com/PennyLaneAI/pennylane/pull/6632 but we decided to exclude it for now due to the following complications (and potential solutions):

- The diagonalize_measurements transform assumes that split_non_commuting has been applied (wouldn't make sense otherwise). What if the device requires diagonalization of some observables but supports commuting measurements?
  - We consider this enough of an edge case that a device like that should just implement its own preprocessing transform program.
- The diagonalize_measurements transform raises an error when it sees anything that is not one of PauliX, PauliY, PauliZ, Hadamard, or a linear combination of the four. For example, if a device supports Hermitian but not Hadamard, it would expect that Hermitian is allowed but Hadamard is diagonalized.
  - We might consider changing diagonalize_measurements to simply leave unrecognized observables as is instead of raising an error. The unsupported observables that are not diagonalized will remain in the circuit past the diagonalize_measurements transform but caught later in the transform program by validate_observables.
- Diagonalization produces additional gates that the device may not support.
  - decompose should be applied after diagonalization. Hopefully with restructured decompositions, diagonalizing gates can always be mapped to the device native gate set.

### CODE REVIEW

File: pennylane/transforms/diagonalize_measurements.py [hunk 1]
```diff
@@ -60,6 +60,12 @@ def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs, to
         circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
         can be applied.
 
+    .. note::
+        This transform will diagonalize what it can, i.e., `qml.X`, `qml.Y`, `qml.Z`, `qml.Hadamard`,
+        `qml.Identity`, or a linear combination of them. Any other observable will be skipped and
+        left undiagonalized. In the context of preprocessing a quantum circuit for execution on a
+        device, observable validation should be performed after this transform to ensure that all
+        undiagonalized observables are supported by the device.
```
Comment (isaacdevlugt) [inline pennylane/transforms/diagonalize_measurements.py:68]: I would say something very similar to what you wrote in the PR description here! Something like this:

> The `diagonalize_measurements` transform assumes that `split_non_commuting` has already been applied, and will diagonalize measurements that are `X`, `Y`, `Z`, `Hadamard`, `Identity`, or any linear combination of the five. Any observables that are not recognized by `diagonalize_measurements` will not raise an error, deferring to the device's validation for supported measurements later on.

> Lastly, if `diagonalize_measurements` produces additional gates that the device does not support, the `decompose` transform will be applied to ensure that relevant additional gates are decomposed to those that the device supports.

**Review by isaacdevlugt — CHANGES_REQUESTED** [review]
Hey Astral! Couple questions and comments:

> The diagonalize_measurements transform assumes that split_non_commuting has been applied (wouldn’t make sense otherwise). What if the device requires diagonalization of some observables but supports commuting measurements?

Do you mean to ask “What if the device requires diagonalization of some observables but supports non-commuting measurements? :thinking_bread: 

> The diagonalize_measurements transform raises an error when it sees anything that is not one of PauliX, PauliY, PauliZ, Hadamard, or a linear combination of the four. For example, if a device supports Hermitian but not Hadamard, it would expect that Hermitian is allowed but Hadamard is diagonalized.
> We might consider changing diagonalize_measurements to simply leave unrecognized observables as is instead of raising an error. The unsupported observables that are not diagonalized will remain in the circuit past the diagonalize_measurements transform but caught later in the transform program by validate_observables.

Agree with you here!

> Diagonalization produces additional gates that the device may not support.
> decompose should be applied after diagonalization. Hopefully with restructured decompositions, diagonalizing gates can always be mapped to the device native gate set.

Agreed! I added a suggestion to the docstring.

**Comment by astralcai** [comment]
> Do you mean to ask “What if the device requires diagonalization of some observables but supports non-commuting measurements?

@isaacdevlugt For example, a device supports measuring `X`, `Y`, and `Z` but not `Hadamard`, so it would expect that `diagonalize_measurements` is applied such that `Hadamard` is diagonalized. Since `diagonalize_measurements` requires that all the tapes do not contain non-commuting observables, we have to first apply `split_non_commuting` to prepare the tapes for proper diagonalization. If the device actually supports simultaneously measuring non-commuting observables, such as `X(0)` and `Z(0)` at the same time, let's say a circuit contains measurements of `X(0)` and `Z(0)`, the user might be expecting a single execution, but because we applied `split_non_commuting`, they get two.

File: pennylane/transforms/diagonalize_measurements.py (hunk 1 above)
Comment (astralcai) [reply pennylane/transforms/diagonalize_measurements.py:68]: There is a note above this one that mentions `split_non_commuting`.

Comment (astralcai) [reply pennylane/transforms/diagonalize_measurements.py]: Updated.

File: pennylane/transforms/diagonalize_measurements.py [hunk 2]
```diff
@@ -60,6 +60,14 @@ def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs, to
         circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
         can be applied.
 
+    .. note::
+        This transform will diagonalize what it can, i.e., ``qml.X``, ``qml.Y``, ``qml.Z``,
+        ``qml.Hadamard``, ``qml.Identity``, or a linear combination of them. Any unrecognized
+        observable will not raise an error, deferring to the device's validation for supported
+        measurements later on. Lastly, if ``diagonalize_measurements`` produces additional gates
+        that the device does not support, the :func:`~pennylane.devices.preprocess.decompose``
+        transform should be applied to ensure that the additional gates are decomposed to those
```
Comment (isaacdevlugt) [inline pennylane/transforms/diagonalize_measurements.py:69]:
```suggestion
        transform is applied internally to ensure that the additional gates are decomposed to those
```

**Review by isaacdevlugt — APPROVED** [review]
Thanks Astral!

File: pennylane/transforms/diagonalize_measurements.py (hunk 2 above)
Comment (astralcai) [reply pennylane/transforms/diagonalize_measurements.py:69]: The `decompose` transform is not automatically applied after `diagonalize_measurements`. Well it is if the device is using our default preprocessing transform, but if the device overrides its own `preprocess`, they need to remember to apply `decompose` after `diagonalize_measurements` themselves, so "should" is more appropriate here.

File: pennylane/transforms/diagonalize_measurements.py [hunk 3]
```diff
@@ -60,6 +60,14 @@ def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs, to
         circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
         can be applied.
 
+    .. note::
+        This transform will diagonalize what it can, i.e., ``qml.X``, ``qml.Y``, ``qml.Z``,
+        ``qml.Hadamard``, ``qml.Identity``, or a linear combination of them. Any unrecognized
+        observable will not raise an error, deferring to the device's validation for supported
+        measurements later on. Lastly, if ``diagonalize_measurements`` produces additional gates
+        that the device does not support, the :func:`~pennylane.devices.preprocess.decompose``
```
Comment (lillian542) [inline pennylane/transforms/diagonalize_measurements.py:68]: currently renders strangely
```suggestion
        that the device does not support, the :func:`~pennylane.devices.preprocess.decompose`
```

File: pennylane/transforms/diagonalize_measurements.py [hunk 4]
```diff
@@ -60,6 +60,14 @@ def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs, to
         circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
         can be applied.
 
+    .. note::
```
Comment (lillian542) [inline pennylane/transforms/diagonalize_measurements.py:63]: It looks kind of funny that there's a note right above this one, and then this second note with two separate things inside it. Should we make them either one large note, or three individual notes?

File: pennylane/transforms/diagonalize_measurements.py [hunk 5]
```diff
@@ -60,6 +60,14 @@ def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs, to
...
+        This transform will diagonalize what it can, i.e., ``qml.X``, ``qml.Y``, ``qml.Z``,
+        ``qml.Hadamard``, ``qml.Identity``, or a linear combination of them. Any unrecognized
+        observable will not raise an error, deferring to the device's validation for supported
```
Comment (lillian542) [inline pennylane/transforms/diagonalize_measurements.py:66]: > Any unrecognized observable will not raise an error, deferring to the device's validation for supported measurements later on.

Should we add an example of this behaviour under Usage Details, showing a circuit where some things are diagonalized, and others are left unchanged?

**Review by lillian542 — COMMENTED** [review]
Looks good to me! Just a few docs suggestions, see what you think :)

File: pennylane/transforms/diagonalize_measurements.py (hunk 4 above)
Comment (astralcai) [reply pennylane/transforms/diagonalize_measurements.py]: Updated it to one large note. Let me know if you have a different preference here.

File: pennylane/transforms/diagonalize_measurements.py [hunk 6]
```diff
@@ -60,6 +60,14 @@ def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs, to
...
+        measurements later on. Lastly, if ``diagonalize_measurements`` produces additional gates
+        that the device does not support, the :func:`~pennylane.devices.preprocess.decompose``
+        transform should be applied to ensure that the additional gates are decomposed to those
+        that the device supports.
```
Comment (isaacdevlugt) [inline pennylane/transforms/diagonalize_measurements.py:69]: @astralcai re: above, how about this?

```suggestion
        transform is applied by PennyLane devices by default to ensure that the additional gates are decomposed to those
        that the device supports. 
```

Comment (astralcai) [reply pennylane/transforms/diagonalize_measurements.py:69]: I don't think that's true, because most PennyLane devices such as `default.qubit` don't actually apply the `diagonalize_measurements` transform, because they already support measuring all of `X`, `Y`, `Z`, and `Hadamard`. This note is meant for device developers thinking about adding `diagonalize_measurements` to their own `preprocess` implementation, to remind them that if you choose to include this transform, make sure that your `decompose` is placed after it, not before it. This is automatically the case if the device developer does not override the `preprocess` but simply use the default implementation inherited from the device API, but if they are reading the docs for this transform, they are most likely implementing their own `preprocess`, which means that nothing is done by default for them and they need to remember to do this.

File: pennylane/transforms/diagonalize_measurements.py (hunk 2 above)
Comment (astralcai) [inline pennylane/transforms/diagonalize_measurements.py]:
```suggestion
        measurements later on. 
        
        Lastly, if you are including ``diagonalize_measurements`` as part of the transform program
        constructed in a custom implementation of ``Device.preprocess_transforms`` for your
        plugin device, note that the transform may produce additional gates that the device does 
        not support, so the :func:`~pennylane.devices.preprocess.decompose`` transform should 
        be applied after to ensure that the additional gates are decomposed to those
```
@isaacdevlugt  Is this better?

**Review by albi3ro — APPROVED** [review]
While this surfaces weird enough edge cases and behaviour,  I think these edge cases are obscure enough that they shouldn't cause enough harm.

Potentially we should check the device test suite and add tests for things like this.  Then we insist devices either raise an error in preprocessing, or give us actual results.  But that can be done later.

**Review by lillian542 — APPROVED** [review]
Looks ready to go 🚀

File: pennylane/transforms/diagonalize_measurements.py (hunk 6 above)
Comment (isaacdevlugt) [reply pennylane/transforms/diagonalize_measurements.py:69]: > if they are reading the docs for this transform, they are most likely implementing their own preprocess, which means that nothing is done by default for them and they need to remember to do this.

I don't think we should assume this! I imagine a user being confused as to why their circuit has gates in it that they didn't put there, so they go down the rabbit hole which leads to `diagonalize_measurements`. In this docstring, we should be clear about when `decompose` is applied and, secondarily, when it _should_ be applied (if that makes sense). 

Having said all of this, I've marinated over your feedback and I think it's probably more confusing to try and elaborate on this point than to leave what's currently there. So, I'm fine with the docstring as written 👍.

Comment (isaacdevlugt) [reply pennylane/transforms/diagonalize_measurements.py:69]: Feel free to resolve this thread and go ahead with the PR 🙂. We can always come back to this if it turns out that a user gets confused
